relays/ethereum/src/message_race_loop.rs

```diff
@@ -0,0 +1,337 @@
+							latest_nonce,
+						);
+
+						strategy.source_nonce_updated(at_block, latest_nonce);
+					},
+					&mut source_go_offline_future,
+					|delay| async_std::task::sleep(delay),
+					|| format!("Error retrieving latest nonce from {}", P::source_name()),
+				).fail_if_connection_error(FailedClient::Source)?;
+			},
+			latest_nonce = target_latest_nonce => {
+				target_latest_nonce_required = false;
+
+				target_client_is_online = process_future_result(
+					latest_nonce,
+					&mut target_retry_backoff,
+					|(_, latest_nonce)| {
+						log::debug!(
+							target: "bridge",
+							"Received latest nonce from {}: {:?}",
+							P::target_name(),
+							latest_nonce,
+						);
+
+						strategy.target_nonce_updated(latest_nonce, &mut race_state);
+					},
+					&mut target_go_offline_future,
+					|delay| async_std::task::sleep(delay),
+					|| format!("Error retrieving latest nonce from {}", P::target_name()),
+				).fail_if_connection_error(FailedClient::Target)?;
+			},
+
+			// proof generation and submittal
+			proof = source_generate_proof => {
+				source_client_is_online = process_future_result(
+					proof,
+					&mut source_retry_backoff,
+					|(at_block, nonces_range, proof)| {
+						log::debug!(
+							target: "bridge",
+							"Received proof for nonces in range {:?} from {}",
+							nonces_range,
+							P::source_name(),
+						);
+
+						race_state.nonces_to_submit = Some((at_block, nonces_range, proof));
+					},
+					&mut source_go_offline_future,
+					|delay| async_std::task::sleep(delay),
+					|| format!("Error generating proof at {}", P::source_name()),
+				).fail_if_connection_error(FailedClient::Source)?;
+			},
+			proof_submit_result = target_submit_proof => {
+				target_client_is_online = process_future_result(
+					proof_submit_result,
+					&mut target_retry_backoff,
+					|nonces_range| {
+						log::debug!(
+							target: "bridge",
+							"Successfully submitted proof of nonces {:?} to {}",
+							nonces_range,
+							P::target_name(),
+						);
+
+						race_state.nonces_to_submit = None;
+						race_state.nonces_submitted = Some(nonces_range);
+					},
+					&mut target_go_offline_future,
+					|delay| async_std::task::sleep(delay),
+					|| format!("Error submitting proof {}", P::target_name()),
+				).fail_if_connection_error(FailedClient::Target)?;
+			}
+		}
+
+		if stall_countdown.elapsed() > stall_timeout {
+			return Err(FailedClient::Both);
+		} else if race_state.nonces_to_submit.is_none() && race_state.nonces_submitted.is_none() && strategy.is_empty()
+		{
+			stall_countdown = Instant::now();
+		}
+
+		if source_client_is_online {
+			source_client_is_online = false;
+
+			if let Some(nonces_range) = strategy.select_nonces_to_deliver(&race_state) {
+				log::debug!(
+					target: "bridge",
+					"Asking {} to prove nonces in range {:?}",
+					P::source_name(),
+					nonces_range,
+				);
+				let at_block = race_state.source_state.as_ref().expect("TODO").best_self.clone();
```

Comment:
TODO?

Reply:
:) Sorry- should have checked for TODOs left